classes/steps/lookups/cohort_lookup_step.php

```diff
@@ -0,0 +1,148 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_trigger\steps\lookups;
+
+defined('MOODLE_INTERNAL') || die;
+require_once($CFG->dirroot  . '/cohort/lib.php');
+
+/**
+ * A lookup step that takes a user's ID and returns a string of
+ * all the cohorts that the user is currently assigned to
+ *
+ * @package    tool_trigger
+ * @author     Paul Damiani <Paul Damiani@catalyst-au.net>
+ * @copyright  Catalyst IT, 2018
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class cohort_lookup_step extends base_lookup_step
+{
+
+    use \tool_trigger\helper\datafield_manager;
+
+    /**
+     * The data field to get the cohort id from.
+     * @var string
+     */
+    private $cohortidfield = null;
+
+    /**
+     * The prefix to put before the new fields added to the workflow data.
+     *
+     * @var string
+     */
+    private $outputprefix = null;
+
+    /**
+     * The fields suplied by this step.
+     * A string containing all the cohorts a user is assigned to.
+     *
+     * @var array
+     */
+    private static $stepfields = array(
+        'user_cohorts'
+    );
+
+    protected function init()
+    {
+        $this->useridfield = $this->data['useridfield'];
+        $this->outputprefix = $this->data['outputprefix'];
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \tool_trigger\steps\base\base_step::execute()
+     */
+    public function execute($step, $trigger, $event, $stepresults)
+    {
+        global $DB;
+        $datafields = $this->get_datafields($event, $stepresults);
+
+        if (!array_key_exists($this->useridfield, $datafields)) {
+            throw new \invalid_parameter_exception("Specified userid field not present in the workflow data: "
+                . $this->useridfield);
+        }
+
+        $cohortdata = cohort_get_user_cohorts($datafields[$this->useridfield]);
```

Comment:
this looks like a recent api addition, so will mean we can't use this plugin in older moodles. Just make sure travis matrix is current

Reply:
... or, if we want to keep support for old moodles then we can't use this, which is annoying but probably the best approach

Reply:
So back to SQL? @brendanheywood